pkg/ddl/index.go

```diff
@@ -599,6 +671,282 @@ func decodeAddIndexArgs(job *model.Job) (
+			// We already have a column with the same column name.
+			// TODO: refine the error message
+			return nil, infoschema.ErrColumnExists.GenWithStackByArgs(hiddenCol.Name)
+		}
+	}
+
+	if len(hiddenCols) > 0 {
+		for _, hiddenCol := range hiddenCols {
+			InitAndAddColumnToTable(tblInfo, hiddenCol)
+		}
+	}
+	if err = checkAddColumnTooManyColumns(len(tblInfo.Columns)); err != nil {
+		return nil, errors.Trace(err)
+	}
+	indexInfo, err = BuildIndexInfo(
+		nil,
+		tblInfo,
+		indexName,
+		isPK,
+		unique,
+		isVector,
+		indexPartSpecifications,
+		indexOption,
+		model.StateNone,
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if isPK {
+		if _, err = CheckPKOnGeneratedColumn(tblInfo, indexPartSpecifications); err != nil {
+			return nil, err
+		}
+	}
+	indexInfo.ID = AllocateIndexID(tblInfo)
+	tblInfo.Indices = append(tblInfo.Indices, indexInfo)
+	if err = checkTooManyIndexes(tblInfo.Indices); err != nil {
+		return nil, errors.Trace(err)
+	}
+	// Here we need do this check before set state to `DeleteOnly`,
+	// because if hidden columns has been set to `DeleteOnly`,
+	// the `DeleteOnly` columns are missing when we do this check.
+	if err := checkInvisibleIndexOnPK(tblInfo); err != nil {
+		return nil, err
+	}
+	logutil.DDLLogger().Info("[ddl] run add index job", zap.String("job", job.String()), zap.Reflect("indexInfo", indexInfo))
+	return indexInfo, nil
+}
+
+func (w *worker) onCreateVectorIndex(jobCtx *jobContext, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	// Handle the rolling back job.
+	if job.IsRollingback() {
+		ver, err = onDropIndex(jobCtx, t, job)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		return ver, nil
+	}
+
+	// Handle normal job.
+	schemaID := job.SchemaID
+	tblInfo, err := GetTableInfoAndCancelFaultJob(t, job, schemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	if err := checkTableTypeForVectorIndex(tblInfo); err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	var (
+		indexName              pmodel.CIStr
+		indexOption            *ast.IndexOption
+		indexPartSpecification *ast.IndexPartSpecification
+		funcExpr               string
+	)
+	err = job.DecodeArgs(&indexName, &indexPartSpecification, &indexOption, &funcExpr)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	indexPartSpecification.Expr, err = generatedexpr.ParseExpression(funcExpr)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	defer func() {
+		indexPartSpecification.Expr = nil
+	}()
+
+	indexInfo, err := checkAndBuildIndexInfo(job, tblInfo, indexName, false, false, true, []*ast.IndexPartSpecification{indexPartSpecification}, indexOption, nil)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	originalState := indexInfo.State
+	switch indexInfo.State {
+	case model.StateNone:
+		// none -> delete only
+		indexInfo.State = model.StateDeleteOnly
+		moveAndUpdateHiddenColumnsToPublic(tblInfo, indexInfo)
```

Comment:
If `indexInfo` is guarenteed to have no hidden columns, this function is not necessary.